src/coordtest/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+                            .with_sc(|sc| Box::pin(async move { sql(sc, query).await }))
+                            .await;
+                        let mut failed = Ok(());
+                        match results {
+                            Ok(result) => {
+                                for r in result {
+                                    match r {
+                                        ExecuteResponse::SendingRows(rows) => match rows.await {
+                                            PeekResponse::Rows(rows) => {
+                                                for row in rows {
+                                                    for col in row.iter() {
+                                                        if col != Datum::True {
+                                                            failed = Err(anyhow!("datum != true"));
+                                                        }
+                                                    }
+                                                }
+                                            }
+                                            r => {
+                                                failed = Err(anyhow!("{:?}", r));
+                                            }
+                                        },
+                                        _ => panic!("expected SendingRows"),
+                                    };
+                                }
+                            }
+                            Err(err) => {
+                                failed = Err(err);
+                            }
+                        };
+                        match failed {
+                            Ok(_) => {
+                                break;
+                            }
+                            Err(err) => {
+                                if start.elapsed() > Duration::from_secs(5) {
+                                    panic!("{}", err);
+                                }
+                                // Bump the timestamp. This is necessary because sources ingest at varying
+                                // rates and we need to allow sinces to move forward so we can see new data.
+                                let mut ts = ct.timestamp.lock().unwrap();
+                                *ts += 1;
+                            }
+                        }
+                    }
+                    "".into()
+                }
+                "update-upper" => {
+                    let catalog = ct.make_catalog().await;
+                    let mut updates = vec![];
+                    for line in tc.input.lines() {
+                        let mut line = line.split_whitespace();
+                        let id = catalog.get(line.next().unwrap());
+                        let ts: Timestamp = line.next().unwrap().parse().unwrap();
+                        assert!(line.next().is_none());
+                        // A ts <= 1 won't do anything.
```

Comment:
why won't 1 do anything? is it because the only available time to send answers at is 0 and we refuse to do that?

Reply:
"Anything" was wrong. It'll advance the upper, but not the since which uses `upper-1`, and is often the desired side effect of running update-upper. Reworded.